### PR TITLE
Computing, storing, and restoring conservative call graphs with LLVM

### PR DESCRIPTION
**Pull request by Prabhuk**
Reviving the original patchset: https://reviews.llvm.org/D105916

### CODE REVIEW

**Comment by ilovepi** [comment]
while its kind of nice to see the complete patch, it looks to me like each of these changes is independent, and should be stacked rather than commited in bulk. Many of us have had good experience with `spr`, so I'd suggest maybe abandoning this PR and re-uploading a stack that can have each part reviewed independently. This will make our review much easier, and the independent changes can land on their own. Plus it will make adding reviewers much easier.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -2714,7 +2714,17 @@ static void setLinkageForGV(llvm::GlobalValue *GV, const NamedDecl *ND) {
 
 void CodeGenModule::CreateFunctionTypeMetadataForIcall(const FunctionDecl *FD,
                                                        llvm::Function *F) {
-  // Only if we are checking indirect calls.
+  bool EmittedMDIdGeneralized = false;
+  if (CodeGenOpts.CallGraphSection &&
+      (!F->hasLocalLinkage() ||
+       F->getFunction().hasAddressTaken(nullptr, /* IgnoreCallbackUses */ true,
+                                        /* IgnoreAssumeLikeCalls */ true,
+                                        /* IgnoreLLVMUsed */ false))) {
```
Comment (arsenm) [inline clang/lib/CodeGen/CodeGenModule.cpp:2722]:
```suggestion
       F->getFunction().hasAddressTaken(nullptr, /* IgnoreCallbackUses= */ true,
                                        /* IgnoreAssumeLikeCalls= */ true,
                                        /* IgnoreLLVMUsed= */ false))) {
```

**Comment by Prabhuk** [comment]
Thanks for the comments. I am working on breaking this into smaller PRs. I'll send them out as soon as possible.

**Comment by Prabhuk** [comment]
Closing this pull request as I have sent out stacked pull requests to land the patches individually.

**Pull request closed without merging**